### PR TITLE
feat:[CDS-79814]: optimized exportVariables functor

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- optimized ExportedVariables functor to directly resolve using sweeping output rather than rendering the output expression again.
- A utility to support resolving (resolve and resolveOptional) sweeping outputs without expression filtered by groupName (in Refobject) is added.
- A bug fix: OutputAlias input variable expression fixed by adding uuid field
- Testing from utility POV: resolving variables at pipeline, stage, stepGroup scope. Also tested nested stepGroups to return instance of most near parent.
- Testing from exportVariables POV: [Test Plan](https://harness.atlassian.net/wiki/spaces/PIPE/pages/21482898438/Tech+Spec+Export+Shell+Script+Variables+to+different+Scopes#Test-Cases)

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)

### CODE REVIEW

File: pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/OutputAlias.java [hunk 1]
```diff
@@ -32,6 +35,10 @@
 @FieldDefaults(level = AccessLevel.PRIVATE)
 @TypeAlias("outputAlias")
 public class OutputAlias {
+  @JsonProperty(YamlNode.UUID_FIELD_NAME)
```
Comment (naman-goenka) [inline pipeline-service/modules/orchestration-steps/contracts/src/main/java/io/harness/steps/shellscript/OutputAlias.java:38]: Reason: For `<+pipeline.stages.test.spec.execution.steps.ShellScript_1.spec.outputAlias.key>` variable creation was failing without uuid. No chema changes should be required here

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 7 of 12 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 12 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java` line 90 at r2](https://reviewable.io/reviews/harness/harness-core/53562#-NfW-gJM8gknquCuP6fN:-NfW-gJM8gknquCuP6fO:bcjcvpq) ([raw file](https://github.com/harness/harness-core/blob/89702e7d2a47c686fc2950874521bf299c307daa/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java#L90)):*
> ```Java
>   public String resolveUsingLevelRuntimeIdx(String planExecutionId, List<String> levelRuntimeIdx, RefObject refObject) {
>     String name = refObject.getName();
>     // We can't filter by groupName provided in rejObject via this utility
> ```

@sahilhindwani-harness i think we can do this right?

**Comment by naman-goenka** [comment]
trigger messagecheck

**Comment by naman-goenka** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/pms/sdk/core/resolver/RefObjectUtils.java` line 62 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyFrzl4kFFWpMFgbo5:-NfyFrzl4kFFWpMFgbo6:bwm19n0) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/pms/sdk/core/resolver/RefObjectUtils.java#L62)):*
> ```Java
>   }
> 
>   public RefObject getSweepingOutputRefObject(String name, String groupName) {
> ```

nit: can we rename the function to be 
getSweepingOutputRefObjectUsingGroup?

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java` line 90 at r2](https://reviewable.io/reviews/harness/harness-core/53562#-NfW-gJM8gknquCuP6fN:-NfyGkg56clFLj5HlGst:b-kfkm11) ([raw file](https://github.com/harness/harness-core/blob/89702e7d2a47c686fc2950874521bf299c307daa/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java#L90)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@sahilhindwani-harness i think we can do this right?
</blockquote></details>

@archit-harness , we are getting the levelRuntimeIdx directly, I don't think we can filter via group here.

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java` line 54 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyHMsD17EFiqITwIWP:-NfyHMsD17EFiqITwIWQ:bynl8n6) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java#L54)):*
> ```Java
>       return prepareLevelRuntimeIdIndices(ambiance);
>     }
>     if (EmptyPredicate.isEmpty(ambiance.getLevelsList())) {
> ```

can we merge this if with above?
Like if EmptyPredicate.isEmpty(groupName) || EmptyPredicate.isEmpty(ambiance.getLevelsList())

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java` line 48 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyHhm9D4q54F0yDTHX:-NfyHhmA2iFjzksx2LVY:bor6eoi) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java#L48)):*
> ```Java
>   }
> 
>   public List<String> prepareLevelRuntimeIdIndicesUsingGroupName(
> ```

Can we pass the groupName to prepareLevelRuntimeIdIndices? That way code can be simplified.

**Review by sahilhindwani-harness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 12 files at r1, 2 of 2 files at r2, 7 of 7 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 9 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by archit-harness — DISMISSED** [review]
> Reviewed 1 of 12 files at r1, 7 of 7 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 1 change requests. 1 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java` line 160 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyIbwP1fhcY-gn8l_L:-NfyIbwP1fhcY-gn8l_M:b-sla6s7) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java#L160)):*
> ```Java
>       return resolveOptionalUsingRuntimeId(ambiance, refObject);
>     }
>     // TODO(sahil): Add implementation for groupName in refObject for expression-like ref name
> ```

nit: @sahilhindwani-harness what will be this usecase?

**Review by naman-goenka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 1 change requests. 2. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/pms/sdk/core/resolver/RefObjectUtils.java` line 62 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyFrzl4kFFWpMFgbo5:-Nfyb85_5x-P8yTPkEr9:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/pms/sdk/core/resolver/RefObjectUtils.java#L62)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

nit: can we rename the function to be 
getSweepingOutputRefObjectUsingGroup?
</blockquote></details>

done
___
*[`pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java` line 48 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyHhm9D4q54F0yDTHX:-NfybtlC4xjFXI0vg9fz:b-ecm81h) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java#L48)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Can we pass the groupName to prepareLevelRuntimeIdIndices? That way code can be simplified.
</blockquote></details>

prepareLevelRuntimeIdIndices is used at many other places,  doing this to reduce blast radius.
___
*[`pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java` line 54 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyHMsD17EFiqITwIWP:-NfybC436oPWxrA9fUO1:bd93yso) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java#L54)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

can we merge this if with above?
Like if EmptyPredicate.isEmpty(groupName) || EmptyPredicate.isEmpty(ambiance.getLevelsList())
</blockquote></details>

doing this to reduce dependency on prepareLevelRuntimeIdIndices. In the case where the behaviour changes in prepareLevelRuntimeIdIndices.

**Review by naman-goenka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java` line 54 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyHMsD17EFiqITwIWP:-Nfydq2a9SallZaWgJMZ:bc5c6h0) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration-beans/contracts/src/main/java/io/harness/engine/pms/data/ResolverUtils.java#L54)):*
<details><summary><i>Previously, naman-goenka (Naman Goenka) wrote…</i></summary><blockquote>

doing this to reduce dependency on prepareLevelRuntimeIdIndices. In the case where the behaviour changes in prepareLevelRuntimeIdIndices.
</blockquote></details>

reverted to the suggested behaviour

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger ut4

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java` line 160 at r4](https://reviewable.io/reviews/harness/harness-core/53562#-NfyIbwP1fhcY-gn8l_L:-Ng-1nNs-cU1b4cHrjfx:b-j0yot3) ([raw file](https://github.com/harness/harness-core/blob/a196255d4f39db85b83e43ea95a019551ce98907/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsSweepingOutputServiceImpl.java#L160)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

nit: @sahilhindwani-harness what will be this usecase?
</blockquote></details>

Not sure at this moment therefore added a todo.

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53562)* status: 6 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Pull request merged**